design.md

```diff
@@ -0,0 +1,100 @@
+# V1
+
+The current code base I will call "v1" (referred to below as the current
+version as well).
+
+The current version is working fairly well in production (overall data
+transfer for tc artifacts is now about $2-8~ dollars a day when
+transferring in aws). In that sense things are great... There are a
+number of other issues we have though:
+
+
+  - The current design is not very flexible (one source -> dest)
+
+      This problem is very very common... You want a canonical source
+      for s3 where you can easily view what's going on but you don't want to
+      incur thousands of dollars of data transfer costs at scale.
+
+      For us to handle more common cases we need a real api.
+
+  - The current design is not trivially scalable:
+
+      We must deploy in _every_ aws region we care about and reconfigure
+      individual callers to be smart enough to pick the right regions that
+      happen to be configured....
+
+      In the worst case a single node goes bad (we have seen this
+      already!) and the node manually requires cycling.
+
+
+## V2
+
+V2 is a series of major changes to how we do both the uploading and
+proxing of things. To start we break up the service into two parts.
+
+  1. The uploader
+  2. The proxy
+
+The uploader will be fairly dumb and only care about uploading stuff
+it's told to upload. It will exist in one region (and eventually) will
+autoscale.
+
+The proxy will be deployed (most likely on heroku) in a fashion that
+enables us to encode all the region detection logic and will be
+responsible for containing the configs of what paths map to what sources
+and metrics for end-to-end experience.
+
+### Proxy
+
+TODO: More details / maybe a graph.
+
+  - Detect AWS Region if possiblee
+  - Hits s3 directly first to see if object is available
+  - Hits uploader and waits for upload event (via sqs+sns) or timeout
+    before redirecting to source (after one last attempt to see if it's
+    ready on s3)
+
+### Upload in Detail
+
+The upload service (uploader) will provide a very simple API via SQS.
+
+An example message body would be something like this (pretty formatting
+shown):
+
+```js
+{
+  source: '....', // url to the source object (typically an s3 object)
+  path: '', // unique path for the object.
+  region: '', //* region where this object should be uploaded to.
+}
+```
+
+The upload service will contain it's own series of buckets located in
+whatever regions we request copies to be made... These buckets will be
+auto generated and contain some baseline rules (like expiring objects
+after N days). The goal here is to eliminate the need to manually
+configure things... You just deploy and it will take care of managing
+the cache buckets.
+
+
+#### Concurrency
+
+TODO: Should we attempt to ensure multiple instance do not upload the
+same objects?
```

Comment:
We should measure S3 `HEAD` time. These requests are rather expensive, so we might want to store the fact that a file has been replicated (and to which regions) in azure table storage, which is much cheaper and possibly faster than S3 `HEAD`.

We could also use azure table storage to mark that we've requested upload if object to given region (at given timestamp, so it'll be retried 20min later). Sure two instances of the "proxy" can still race and as they both send the SQS message before modifying the azure table entity, we could still get two upload operations. But the likelyhood is much smaller because the critical section is smaller.

If uploader also modified table storage it could ensure that it got the last message to upload, so no two upload processes uploads in parallel. This is done by putting the timestamp (with ms) into the message and the azure table entity...